# Análise Exploratória de Dados com SQL

Análise Exploratória de Dados provenientes de uma empresa de varejo no segmento de vestuário com foco no e-commerce.

In [11]:
# Instalando o pacote que possibilita utilizar SQL no Jupyter Notebook
!pip install ipython-sql

  Created wheel for prettytable: filename=prettytable-0.7.2-py3-none-any.whl size=13714 sha256=9b22c436240cc99a461205f815b491aa4e93f4f2ca1171627bee7556be0e2169
  Stored in directory: c:\users\nicolas\appdata\local\pip\cache\wheels\75\f7\28\77a076f1fa8cbeda61aca712815d04d7a32435f04a26a2dd7b
Successfully built prettytable


In [34]:
# Carregando a extensão do SQL
%load_ext sql

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


In [35]:
# Conectando ao dataset já formatado
%sql sqlite:///sales.db

## Análise do Dataset

In [3]:
# Identificando as tabelas do banco de dados
%%sql
SELECT name FROM sqlite_master WHERE type='table';

 * sqlite:///sales.db
Done.


[('customers',), ('orders',), ('products',), ('sales',)]

In [4]:
# Analisando os atributos/colunas e tipos dos registros de cada tabela
%%sql
PRAGMA TABLE_INFO ('customers')

 * sqlite:///sales.db
Done.


[(0, 'customer_id', 'INTEGER', 0, None, 1),
 (1, 'customer_name', 'TEXT', 0, None, 0),
 (2, 'gender', 'TEXT', 0, None, 0),
 (3, 'age', 'INTEGER', 0, None, 0),
 (4, 'home_address', 'TEXT', 0, None, 0),
 (5, 'zip_code', 'INTEGER', 0, None, 0),
 (6, 'city', 'TEXT', 0, None, 0),
 (7, 'state', 'TEXT', 0, None, 0),
 (8, 'country', 'TEXT', 0, None, 0),
 (9, 'age_bracket', 'TEXT AFTER age', 0, None, 0)]

In [5]:
%%sql
PRAGMA TABLE_INFO ('orders')

 * sqlite:///sales.db
Done.


[(0, 'order_id', 'INTEGER', 0, None, 1),
 (1, 'customer_id', 'INTEGER', 0, None, 0),
 (2, 'payment', 'INTEGER', 0, None, 0),
 (3, 'delivery_date', 'date', 0, None, 0),
 (4, 'order_date', 'date', 0, None, 0),
 (5, 'order_date_month', 'date', 0, None, 0)]

In [6]:
%%sql
PRAGMA TABLE_INFO ('products')

 * sqlite:///sales.db
Done.


[(0, 'product_id', 'INTEGER', 0, None, 1),
 (1, 'product_type', 'TEXT', 0, None, 0),
 (2, 'product_name', 'TEXT', 0, None, 0),
 (3, 'size', 'TEXT', 0, None, 0),
 (4, 'colour', 'TEXT', 0, None, 0),
 (5, 'price', 'INTEGER', 0, None, 0),
 (6, 'quantity', 'INTEGER', 0, None, 0),
 (7, 'description', 'TEXT', 0, None, 0)]

In [7]:
%%sql
PRAGMA TABLE_INFO ('sales')

 * sqlite:///sales.db
Done.


[(0, 'sales_id', 'INTEGER', 0, None, 1),
 (1, 'order_id', 'INTEGER', 0, None, 0),
 (2, 'product_id', 'INTEGER', 0, None, 0),
 (3, 'price_per_unit', 'INTEGER', 0, None, 0),
 (4, 'quantity', 'INTEGER', 0, None, 0),
 (5, 'total_price', 'INTEGER', 0, None, 0)]

## Análises Iniciais

### Clientes

In [34]:
# Número de clientes cadastrados
%%sql
SELECT COUNT(customer_id) AS 'Total Customers' 
FROM customers;

 * sqlite:///sales.db
Done.


[(1000,)]

In [6]:
# Número de clientes por gênero
%%sql
SELECT gender AS 'Gender',
COUNT(customer_id) AS 'Customers per Gender' 
FROM customers 
GROUP BY gender
ORDER BY COUNT(customer_id) DESC;

 * sqlite:///sales.db
Done.


[('Male', 143),
 ('Non-binary', 131),
 ('Polygender', 128),
 ('Genderqueer', 127),
 ('Genderfluid', 122),
 ('Bigender', 120),
 ('Female', 115),
 ('Agender', 114)]

In [7]:
# Idade média dos clientes
%%sql
SELECT ROUND(AVG(age)) AS 'Average Age' 
FROM customers;

 * sqlite:///sales.db
Done.


[(50.0,)]

In [8]:
# País onde residem os clientes
%%sql
SELECT country AS 'Country', 
COUNT(customer_id) AS 'Customers per Country' 
FROM customers 
GROUP BY country;

 * sqlite:///sales.db
Done.


[('Australia', 1000)]

In [10]:
# Estados onde residem os clientes
%%sql
SELECT state AS 'State', 
COUNT(customer_id) AS 'Customers per State' 
FROM customers 
GROUP BY state
ORDER BY COUNT(customer_id) DESC;

 * sqlite:///sales.db
Done.


[('South Australia', 139),
 ('Queensland', 134),
 ('New South Wales', 132),
 ('Northern Territory', 125),
 ('Western Australia', 124),
 ('Victoria', 121),
 ('Australian Capital Territory', 121),
 ('Tasmania', 104)]

In [15]:
# Número de cidades distintas onde residem os clientes
%%sql
SELECT COUNT(DISTINCT(city)) AS 'Numbers of Cities' 
FROM customers;

 * sqlite:///sales.db
Done.


[(961,)]

In [37]:
# As 5 cidades onde residem mais clientes
%%sql
SELECT city AS 'City', 
COUNT(customer_id) AS 'Customers per City' 
FROM customers 
GROUP BY city 
ORDER BY COUNT(customer_id) DESC LIMIT 5;

 * sqlite:///sales.db
Done.


[('New Ava', 3),
 ('East Sophia', 3),
 ('East Aidan', 3),
 ('Zacharybury', 2),
 ('Williammouth', 2)]

### Produtos

In [16]:
# Número de tipos de produto cadastrados
%%sql
SELECT product_type AS 'Product Type', 
COUNT(product_id) AS 'Number of Products' 
FROM products 
GROUP BY product_type;

 * sqlite:///sales.db
Done.


[('Jacket', 420), ('Shirt', 420), ('Trousers', 420)]

In [17]:
# Nomes e número dos produtos cadastrados 
%%sql
SELECT product_name AS 'Product Name', 
COUNT(product_id) AS 'Number of Products' 
FROM products 
GROUP BY product_name;

 * sqlite:///sales.db
Done.


[('Bomber', 35),
 ('Camp Collared', 35),
 ('Cardigan', 35),
 ('Cargo Pants', 35),
 ('Casual Slim Fit', 35),
 ('Chambray', 35),
 ('Chinos', 35),
 ('Coach', 35),
 ('Cords', 35),
 ('Cropped', 35),
 ('Cuban Collar', 35),
 ('Denim', 70),
 ('Drawstring', 35),
 ('Dress', 35),
 ('Flannel', 35),
 ('Henley', 35),
 ('High-Waisted', 35),
 ('Joggers', 35),
 ('Leather', 35),
 ('Linen', 35),
 ('Mandarin Collar', 35),
 ('Oxford Cloth', 35),
 ('Parka', 35),
 ('Peacoat', 35),
 ('Pleated', 35),
 ('Polo', 35),
 ('Puffer', 35),
 ('Pullover', 35),
 ('Relaxed Leg', 35),
 ('Shearling', 35),
 ('Slim-Fit', 35),
 ('Tracksuit Bottoms', 35),
 ('Trench Coat', 35),
 ('Windbreaker', 35),
 ('Wool', 35)]

In [21]:
# Os 10 produtos diferentes mais caros
%%sql
SELECT DISTINCT(product_name) AS 'Product Name', 
product_type AS 'Product Type', 
price AS 'Price' 
FROM products 
ORDER BY price DESC LIMIT 10;

 * sqlite:///sales.db
Done.


[('Dress', 'Shirt', 119),
 ('Parka', 'Jacket', 119),
 ('Slim-Fit', 'Trousers', 119),
 ('Cardigan', 'Jacket', 118),
 ('Polo', 'Shirt', 117),
 ('Linen', 'Shirt', 116),
 ('Denim', 'Shirt', 115),
 ('Coach', 'Jacket', 115),
 ('Oxford Cloth', 'Shirt', 114),
 ('Pullover', 'Jacket', 114)]

### Pedidos

In [22]:
# A data de pedido mais antiga e mais recente
%%sql
SELECT MIN(order_date) AS 'Oldest Order Date', 
MAX(order_date) AS 'Recent Order Date' 
FROM orders;

 * sqlite:///sales.db
Done.


[('2021-01-01', '2021-10-24')]

### Vendas

In [6]:
# O número total de vendas (quantidade e receita) e o valor médio de venda
%%sql
SELECT COUNT(sales_id) AS 'Number of Sales', 
SUM(total_price) AS 'Sales Total',
AVG(total_price) AS 'Average Sale Price'
FROM sales;

 * sqlite:///sales.db
Done.


[(5000, 1031800, 206.36)]

In [24]:
# As 10 maiores vendas cadastradas
%%sql
SELECT DISTINCT(order_id) AS 'Number of Order', 
total_price AS 'Sale Price' 
FROM sales 
ORDER BY total_price DESC LIMIT 10;

 * sqlite:///sales.db
Done.


[(45, 357),
 (61, 357),
 (75, 357),
 (77, 357),
 (78, 357),
 (95, 357),
 (100, 357),
 (108, 357),
 (120, 357),
 (123, 357)]

In [25]:
# A venda mais barata e a venda mais cara
%%sql
SELECT MIN(total_price) AS 'Min. Sale Price', 
MAX(total_price) AS 'Max. Sale Price' 
FROM sales;

 * sqlite:///sales.db
Done.


[(90, 357)]

## Perguntas de Negócio

### 1. Qual a faixa etária que mais comprou? (Número de Pedidos e Receita)

In [8]:
%%sql
SELECT c.age AS 'Age', 
COUNT(order_id) AS 'Number of Orders' 
FROM customers AS c 
JOIN orders AS o ON c.customer_id = o.customer_id 
GROUP BY c.age 
ORDER BY COUNT(order_id) 
DESC LIMIT 10;

 * sqlite:///sales.db
Done.


[(57, 34),
 (75, 27),
 (32, 27),
 (63, 26),
 (38, 26),
 (34, 26),
 (45, 24),
 (25, 24),
 (27, 23),
 (59, 22)]

Temos muitas idades diferentes cadastradas, de modo que faz mais sentido responder essa pergunta agrupando as idades em faixas etárias. 

Para tanto, vamos adicionar uma nova coluna/atributo de faixa etária.

In [ ]:
%%sql
ALTER TABLE customers ADD age_bracket TEXT;

In [ ]:
%%sql
UPDATE customers SET age_bracket = 'Youth' WHERE age <= '25';

In [ ]:
%%sql
UPDATE customers SET age_bracket = 'Adult' WHERE age > '25' AND age <= '59;

In [ ]:
%%sql
UPDATE customers SET age_bracket = 'Old' WHERE age >= '60';

In [9]:
%%sql
SELECT age_bracket AS 'Age Bracket', 
COUNT(customer_id) AS 'Number of Customers' 
FROM customers GROUP BY age_bracket;

 * sqlite:///sales.db
Done.


[('Adult', 552), ('Old', 343), ('Youth', 105)]

Agrupando as idades em três faixas etárias, fica mais fácil analisar qual a faixa etária que mais comprou produtos da empresa.

In [12]:
%%sql
SELECT c.age_bracket AS 'Age Bracket', 
COUNT(o.order_id) AS 'Number of Orders', 
SUM(s.total_price) AS 'Total Sales', 
printf('%2.2f%%', (SUM(s.total_price) * 100.00 / (SELECT SUM(s.total_price) FROM sales AS s))) AS '% of Sales' 
FROM customers AS c 
JOIN orders AS o ON c.customer_id = o.customer_id 
JOIN sales AS s ON o.order_id = s.order_id 
GROUP BY c.age_bracket;

 * sqlite:///sales.db
Done.


[('Adult', 2835, 586627, '56.85%'),
 ('Old', 1577, 326381, '31.63%'),
 ('Youth', 588, 118792, '11.51%')]

Podemos ver que 56.85% das vendas são feitas para Adultos (idade entre 25 e 59 anos), 31.63% para Idosos (idade acima de 60 anos), e 11.51% para Jovens (idade abaixo de 24 anos).

### 2. Qual o gênero que mais comprou? (Número de Pedidos e Receita)

In [28]:
%%sql
SELECT c.gender AS 'Gender', 
COUNT(o.order_id) AS 'Number of Orders', 
SUM(s.total_price) AS 'Total Sales', 
printf('%2.2f%%', (SUM(s.total_price) * 100.00 / (SELECT SUM(s.total_price) FROM sales AS s))) AS '% of Sales' 
FROM customers AS c 
JOIN orders AS o ON c.customer_id = o.customer_id 
JOIN sales AS s ON o.order_id = s.order_id 
GROUP BY c.gender 
ORDER BY SUM(s.total_price) DESC;

 * sqlite:///sales.db
Done.


[('Female', 712, 145404, '14.09%'),
 ('Genderfluid', 687, 142077, '13.77%'),
 ('Male', 672, 138988, '13.47%'),
 ('Polygender', 636, 132548, '12.85%'),
 ('Genderqueer', 647, 132389, '12.83%'),
 ('Non-binary', 598, 126201, '12.23%'),
 ('Agender', 526, 107179, '10.39%'),
 ('Bigender', 522, 107014, '10.37%')]

Tratando de gêneros, podemos ver que não existe uma discrepância muito grande entre as vendas, mas são as mulheres que mais compram, representando 14.09% das vendas.

### 3. Qual o Estado que mais comprou? (Número de Pedidos e Receita)

In [29]:
%%sql
SELECT c.state AS 'State', 
COUNT(o.order_id) AS 'Number of Orders', 
SUM(s.total_price) AS 'Total Sales',
printf('%2.2f%%', (SUM(s.total_price) * 100.00 / (SELECT SUM(s.total_price) FROM sales AS s))) AS '% of Sales' 
FROM customers AS c 
JOIN orders AS o ON c.customer_id = o.customer_id 
JOIN sales AS s ON o.order_id = s.order_id 
GROUP BY c.state 
ORDER BY SUM(s.total_price) DESC;

 * sqlite:///sales.db
Done.


[('South Australia', 718, 147816, '14.33%'),
 ('Queensland', 687, 142062, '13.77%'),
 ('New South Wales', 647, 131744, '12.77%'),
 ('Australian Capital Territory', 638, 130442, '12.64%'),
 ('Western Australia', 604, 125430, '12.16%'),
 ('Tasmania', 586, 122095, '11.83%'),
 ('Victoria', 591, 120357, '11.66%'),
 ('Northern Territory', 529, 111854, '10.84%')]

Da mesma forma que os gêneros, não temos uma discrepância muito grande entre as vendas, sendo o Estado que mais comprou o Estado de South Australia, representando 14.33% das vendas.

### 4. Quais foram os meses com o maior número de pedidos? (Número de Pedidos e Receita)

In [31]:
%%sql
SELECT o.order_date_month AS 'Month of Order',
COUNT(o.order_id) AS 'Number of Orders', 
SUM(s.total_price) AS 'Total Sales', 
printf('%2.2f%%', (SUM(s.total_price) * 100.00 / (SELECT SUM(s.total_price) FROM sales AS s))) AS '% of Sales' 
FROM orders AS o 
JOIN sales AS s ON o.order_id = s.order_id 
GROUP BY o.order_date_month 
ORDER BY SUM(s.total_price) DESC;

 * sqlite:///sales.db
Done.


[(3, 635, 131364, '12.73%'),
 (1, 577, 119333, '11.57%'),
 (7, 567, 116081, '11.25%'),
 (8, 505, 104565, '10.13%'),
 (6, 501, 101458, '9.83%'),
 (4, 478, 97530, '9.45%'),
 (9, 458, 96526, '9.36%'),
 (2, 467, 95080, '9.21%'),
 (5, 417, 85597, '8.30%'),
 (10, 395, 84266, '8.17%')]

Considerando que o conjunto de dados possui registros de Janeiro a Outubro, podemos ver que os três melhores meses de venda (até agora) foram Março, Janeiro e Julho. 

### 5. Quais os produtos mais vendidos? (Número de Pedidos e Receita)

In [20]:
%%sql
SELECT p.product_name AS 'Product Name', 
p.product_type AS 'Product Type', 
COUNT(o.order_id) AS 'Number of Orders', 
SUM(s.total_price) AS 'Total Sales', 
printf('%2.2f%%', (SUM(s.total_price) * 100.00 / (SELECT SUM(s.total_price) FROM sales AS s))) AS '% of Sales' 
FROM products AS p 
JOIN sales AS s ON p.product_id = s.product_id 
JOIN orders AS o ON s.order_id = o.order_id 
GROUP BY p.product_name 
ORDER BY SUM(s.total_price) DESC;

 * sqlite:///sales.db
Done.


[('Denim', 'Jacket', 273, 52399, '5.08%'),
 ('Casual Slim Fit', 'Shirt', 154, 36414, '3.53%'),
 ('Trench Coat', 'Jacket', 146, 35581, '3.45%'),
 ('Shearling', 'Jacket', 150, 35334, '3.42%'),
 ('Puffer', 'Jacket', 140, 35164, '3.41%'),
 ('Flannel', 'Shirt', 141, 33158, '3.21%'),
 ('Cropped', 'Trousers', 135, 32660, '3.17%'),
 ('Pleated', 'Trousers', 147, 32340, '3.13%'),
 ('Joggers', 'Trousers', 164, 31062, '3.01%'),
 ('Chambray', 'Shirt', 141, 30740, '2.98%'),
 ('Parka', 'Jacket', 134, 30397, '2.95%'),
 ('Bomber', 'Jacket', 150, 29700, '2.88%'),
 ('Chinos', 'Trousers', 135, 29380, '2.85%'),
 ('Windbreaker', 'Jacket', 143, 29274, '2.84%'),
 ('Cardigan', 'Jacket', 134, 29120, '2.82%'),
 ('Peacoat', 'Jacket', 132, 28994, '2.81%'),
 ('Tracksuit Bottoms', 'Trousers', 130, 28784, '2.79%'),
 ('Slim-Fit', 'Trousers', 154, 27936, '2.71%'),
 ('Oxford Cloth', 'Shirt', 146, 27744, '2.69%'),
 ('Drawstring', 'Trousers', 147, 27648, '2.68%'),
 ('High-Waisted', 'Trousers', 140, 27522, '2.67%'),
 ('Camp Collared', 'Shirt', 139, 27342, '2.65%'),
 ('Cargo Pants', 'Trousers', 134, 27234, '2.64%'),
 ('Relaxed Leg', 'Trousers', 146, 27090, '2.63%'),
 ('Cuban Collar', 'Shirt', 122, 26670, '2.58%'),
 ('Henley', 'Shirt', 146, 26414, '2.56%'),
 ('Wool', 'Trousers', 125, 26394, '2.56%'),
 ('Leather', 'Jacket', 136, 25774, '2.50%'),
 ('Mandarin Collar', 'Shirt', 124, 25724, '2.49%'),
 ('Dress', 'Shirt', 127, 25272, '2.45%'),
 ('Pullover', 'Jacket', 143, 25116, '2.43%'),
 ('Polo', 'Shirt', 131, 25080, '2.43%'),
 ('Coach', 'Jacket', 134, 24645, '2.39%'),
 ('Linen', 'Shirt', 131, 24570, '2.38%'),
 ('Cords', 'Trousers', 126, 23124, '2.24%')]

O produto mais vendido no período foi a Jaqueta de Denim, representado 5.08% das vendas da empresa.

Em seguida, temos a Camisa Slim Fit, a Jaqueta Trench, a Jaqueta Shearling e a Jaqueta Puffer. 

Esses 05 produtos, somados, representam 18.89%¨das vendas da empresa no período.

### 6. Qual o tempo médio entre a data do pedido e a data de entrega?

In [21]:
%%sql
SELECT round(avg(julianday(delivery_date) - julianday(order_date))) AS 'Average Number of Days to Deliver' 
FROM orders; 

 * sqlite:///sales.db
Done.


[(14.0,)]

O número médio de dias entre o pedido e a entrega do produto para o cliente é de 14 dias.

### 7. Qual o desempenho de vendas por trimestre? (Número de Pedidos e Receita)

Como não temos o dado em que trimestre o pedido foi feito, vamos adicionar o atributo/coluna do trimestre no conjunto de dados para facilitar a análise e responder ao questionamento.

In [22]:
%%sql
ALTER TABLE orders ADD order_date_quarter TEXT;

 * sqlite:///sales.db
Done.


[]

In [23]:
%%sql
UPDATE orders SET order_date_quarter = 'Q1' WHERE order_date_month >= 1 AND order_date_month <= 3;

 * sqlite:///sales.db
321 rows affected.


[]

In [24]:
%%sql
UPDATE orders SET order_date_quarter = 'Q2' WHERE order_date_month >= 4 AND order_date_month <= 6;

 * sqlite:///sales.db
288 rows affected.


[]

In [25]:
%%sql
UPDATE orders SET order_date_quarter = 'Q3' WHERE order_date_month >= 7 AND order_date_month <= 9;

 * sqlite:///sales.db
310 rows affected.


[]

In [26]:
%%sql
UPDATE orders SET order_date_quarter = 'Q4' WHERE order_date_month >= 10 AND order_date_month <= 12;

 * sqlite:///sales.db
81 rows affected.


[]

In [27]:
%%sql
SELECT o.order_date_quarter AS 'Quarter', 
COUNT(o.order_id) AS 'Number of Orders', 
SUM(s.total_price) AS 'Sales Total', 
printf('%2.2f%%', (SUM(s.total_price) * 100.00 / (SELECT SUM(s.total_price) FROM sales AS s))) AS '% of Sales' 
FROM orders AS o 
JOIN sales AS s ON o.order_id = s.order_id 
GROUP BY o.order_date_quarter;

 * sqlite:///sales.db
Done.


[('Q1', 1679, 345777, '33.51%'),
 ('Q2', 1396, 284585, '27.58%'),
 ('Q3', 1530, 317172, '30.74%'),
 ('Q4', 395, 84266, '8.17%')]

O Primeiro Trimestre (Janeiro a Março) foi o melhor período de vendas para a empresa.

Todavia, cabe ressaltar que o Quarto Trimestre tem somente as vendas do mês de Outubro, sendo esse o último mês que temos registros no conjunto de dados.